File: FFmpegNative/Native/FormatContext.cpp
#include "FormatContext.h"

#include <cassert>
#include <format>
#include <stdexcept>

#include "FFmpegException.h"
#include "Helper.h"

FormatContext::FormatContext(FormatContext &&other) noexcept
{
    this->raw_fmt_ctx = other.raw_fmt_ctx;
    other.raw_fmt_ctx = nullptr;
}

FormatContext &FormatContext::operator=(FormatContext &&other) noexcept
{
    assert(this != &other);
    this->close();
    this->raw_fmt_ctx = other.raw_fmt_ctx;
    other.raw_fmt_ctx = nullptr;
    return *this;
}

FormatContext::~FormatContext()
{
    this->close();
}

void FormatContext::open_in(std::string_view filename)
{
    assert(this->is_opened() == false);
    AVFormatContext *format_ctx = avformat_alloc_context();
    if (format_ctx == nullptr)
    {
        throw FFmpegException("Failed to allocate format context");
    }
    int ret = avformat_open_input(&format_ctx, filename.data(), nullptr, nullptr);
    if (ret < 0)
    {
        avformat_free_context(format_ctx);
        throw FFmpegException(ret, "Failed to open input file");
    }
    ret = avformat_find_stream_info(format_ctx, nullptr);
    if (ret < 0)
    {
        avformat_close_input(&format_ctx);
        throw FFmpegException(ret, "Failed to find stream info");
    }
    this->raw_fmt_ctx = format_ctx;
}

void FormatContext::open_out(std::string_view filename, std::string_view format)
{
    assert(this->is_opened() == false);
    const AVOutputFormat *output_format = av_guess_format(format.data(), nullptr, nullptr);
    int ret = avformat_alloc_output_context2(&this->raw_fmt_ctx, output_format, nullptr, filename.data());
    if (ret < 0)
    {
        throw FFmpegException(ret, "Failed to allocate output context");
    }
    if ((this->raw_fmt_ctx->oformat->flags & AVFMT_NOFILE) == 0)
    {
        int ret = avio_open(&this->raw_fmt_ctx->pb, filename.data(), AVIO_FLAG_WRITE);
        if (ret < 0)
        {
            avformat_free_context(this->raw_fmt_ctx);
            this->raw_fmt_ctx = nullptr;
            throw FFmpegException(ret, "Failed to open output file");
        }
    }
}

void FormatContext::close()
{
    avformat_close_input(&this->raw_fmt_ctx);
}

const AVFormatContext *FormatContext::get() const
{
    return this->raw_fmt_ctx;
}

AVFormatContext *FormatContext::get()
{
    return this->raw_fmt_ctx;
}

int FormatContext::stream_count() const
{
    assert(this->raw_fmt_ctx != nullptr);
    return this->raw_fmt_ctx->nb_streams;
}

const AVStream *FormatContext::stream(int index) const
{
    assert(this->raw_fmt_ctx != nullptr);
    return this->raw_fmt_ctx->streams[index];
}

AVStream *FormatContext::stream(int index)
{
    assert(this->raw_fmt_ctx != nullptr);
    return this->raw_fmt_ctx->streams[index];
}

AVStream *FormatContext::add_stream(const AVCodecContext *codec)
{
    assert(this->is_opened() == true);
    assert(this->raw_fmt_ctx->oformat != nullptr);

    AVStream *out_stream = avformat_new_stream(this->raw_fmt_ctx, nullptr);
    if (out_stream == nullptr)
    {
        throw FFmpegException("Failed to create output stream");
    }

    int ret = avcodec_parameters_from_context(out_stream->codecpar, codec);
    if (ret < 0)
    {
        throw FFmpegException(ret, "Failed to copy encoder parameters to output stream");
    }

    out_stream->time_base = codec->time_base;
    return out_stream;
}

int FormatContext::best_stream_index(AVMediaType type) const
{
    assert(this->raw_fmt_ctx != nullptr);
    return av_find_best_stream(this->raw_fmt_ctx, type, -1, -1, nullptr, 0);
}

void FormatContext::write_header()
{
    assert(this->is_opened() == true);
    assert(this->raw_fmt_ctx->oformat != nullptr);
    int ret = avformat_write_header(this->raw_fmt_ctx, nullptr);
    if (ret < 0)
    {
        throw FFmpegException(ret, "Failed to write header");
    }
}

void FormatContext::write_trailer()
{
    assert(this->is_opened() == true);
    assert(this->raw_fmt_ctx->oformat != nullptr);
    int ret = av_write_trailer(this->raw_fmt_ctx);
    if (ret < 0)
    {
        throw FFmpegException(ret, "Failed to write trailer");
    }
}

bool FormatContext::is_opened() const
{
    return this->raw_fmt_ctx != nullptr;
}

File: AvaloniaMedia/FFmpeg/Demux/FormatContext.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Runtime.InteropServices;
using AvaloniaMedia.FFmpeg.Model;
using FFmpeg.AutoGen;

namespace AvaloniaMedia.FFmpeg.Demux;

public sealed unsafe class FormatContext : IDisposable
{
    #region AVFormatContext
    public AVFormatContext* FormatContextPointer { get; private set; }
    #endregion

    #region FFmpeg回调的函数委托
    private GCHandle? ioContextHandle;
    private avio_alloc_context_read_packet? readDelegate;
    private avio_alloc_context_seek? seekDelegate;
    private bool ownsAVIO;
    #endregion

    #region Public Methods
    public bool IsOpen => FormatContextPointer != null;
    public double Duration => (double)FormatContextPointer->duration / ffmpeg.AV_TIME_BASE;
    public int StreamCount => (int)FormatContextPointer->nb_streams;
    public AVStream* GetStream(int index) => FormatContextPointer->streams[index];
    public string? Title => GetTitle();
    #endregion

    public string? GetTitle()
    {
        Debug.Assert(IsOpen);

        AVDictionaryEntry* tag = ffmpeg.av_dict_get(FormatContextPointer->metadata, "title", null, 0);
        if (tag == null)
            return null;

        return Marshal.PtrToStringAnsi((IntPtr)tag->value);
    }

    /// <summary>
    /// 打开一个文件作为输入
    /// </summary>
    /// <param name="filename"></param>
    /// <returns></returns>
    /// <exception cref="FFmpegException"></exception>
    public void OpenAsReader(string filename, Func<bool>? interruptCB = null)
    {
        Debug.Assert(IsOpen == false);

        AVFormatContext* formatCtx = ffmpeg.avformat_alloc_context();
        if (formatCtx == null)
        {
            throw new FFmpegException("Failed to allocate format context");
        }

        if (interruptCB != null)
        {
            formatCtx->interrupt_callback = InterruptCallback.Create(interruptCB);
        }

        int result = ffmpeg.avformat_open_input(&formatCtx, filename, null, null);
        if (result < 0)
        {
            ffmpeg.avformat_free_context(formatCtx);
            throw new FFmpegException(result, $"Failed to open input file {filename}");
        }

        result = ffmpeg.avformat_find_stream_info(formatCtx, null);
        if (result < 0)
        {
            ffmpeg.avformat_close_input(&formatCtx);
            throw new FFmpegException(result, $"Failed to find stream info");
        }

        FormatContextPointer = formatCtx;
    }

    public void OpenAsReader(Stream stream)
    {
        Debug.Assert(IsOpen == false);

        // 创建格式上下文
        AVFormatContext* formatCtx = ffmpeg.avformat_alloc_context();
        if (formatCtx == null)
        {
            throw new FFmpegException("Failed to allocate format context");
        }

        // 创建buffer
        const int ioBufferSize = 512 * 1024;
        byte* buffer = (byte*)ffmpeg.av_malloc(ioBufferSize);
        if (buffer == null)
        {
            ffmpeg.avformat_free_context(formatCtx);
            throw new FFmpegException("Failed to allocate AVIO buffer");
        }

        // 创建自定义IO并绑定句柄
        var ioContext = new IOContext(stream);
        var handle = GCHandle.Alloc(ioContext, GCHandleType.Normal);

        if (handle.IsAllocated == false)
        {
            ffmpeg.av_free(buffer);
            ffmpeg.avformat_free_context(formatCtx);
            throw new Exception("Failed to allocate GCHandle for IOContext");
        }

        void* opaque = GCHandle.ToIntPtr(handle).ToPointer();
        var readDelegate = new avio_alloc_context_read_packet(IOContext.FFmpegRead);
        var seekDelegate = new avio_alloc_context_seek(IOContext.FFmpegSeek);

        // 创建AVIO上下文
        AVIOContext* ioCtx = ffmpeg.avio_alloc_context(buffer, ioBufferSize, 0, opaque, readDelegate, null, seekDelegate);
        if (ioCtx == null)
        {
            handle.Free();
            ffmpeg.av_free(buffer);
            ffmpeg.avformat_free_context(formatCtx);
            throw new FFmpegException("Failed to allocate AVIO context");
        }

        // 绑定avio到formatCtx
        formatCtx->pb = ioCtx;

        int result = ffmpeg.avformat_open_input(&formatCtx, null, null, null);
        if (result < 0)
        {
            ffmpeg.avformat_free_context(formatCtx);
            ffmpeg.avio_context_free(&ioCtx);
            handle.Free();
            throw new FFmpegException(result, "Failed to open input file");
        }

        result = ffmpeg.avformat_find_stream_info(formatCtx, null);
        if (result < 0)
        {
            ffmpeg.avformat_close_input(&formatCtx);
            ffmpeg.avio_context_free(&ioCtx);
            handle.Free();
            throw new FFmpegException(result, "Failed to find stream info");
        }

        FormatContextPointer = formatCtx;
        ioContextHandle = handle;
        this.readDelegate = readDelegate;
        this.seekDelegate = seekDelegate;
        ownsAVIO = true;
    }

    public void OpenAsWritter(string fileName, string formatName)
    {
        AVFormatContext* formatCtx = null;
        int result = ffmpeg.avformat_alloc_output_context2(&formatCtx, null, formatName, fileName);
        if (result < 0)
        {
            throw new FFmpegException(result, "Fail to alloc output format context.");
        }
    }

    ~FormatContext() => Dispose();

    public void Close()
    {
        if (FormatContextPointer != null)
        {
            var ctx = FormatContextPointer;
            var interrupt_callback = ctx->interrupt_callback;
            var pb = ctx->pb;
            ffmpeg.avformat_close_input(&ctx);
            FormatContextPointer = null;
            if (ownsAVIO)
                ffmpeg.avio_context_free(&pb);
            InterruptCallback.Release(ref interrupt_callback);
        }
        if (ioContextHandle != null)
        {
            if (ioContextHandle.Value.IsAllocated)
            {
                ioContextHandle.Value.Free();
            }
            ioContextHandle = null;
        }

        readDelegate = null;
        seekDelegate = null;
    }

    private bool disposed = false;
    public void Dispose()
    {
        if (disposed == true)
            return;

        Close();

        GC.SuppressFinalize(this);

        disposed = true;
    }
}